src/System.Memory/src/System/Buffers/ReadOnlySequence.cs

```diff
@@ -128,20 +125,48 @@ public ReadOnlySequence(T[] array, int start, int length)
         /// </summary>
         public ReadOnlySequence(ReadOnlyMemory<T> readOnlyMemory)
         {
-            ReadOnlySequenceSegment segment = new ReadOnlySequenceSegment
+            if (MemoryMarshal.TryGetOwnedMemory(readOnlyMemory, out OwnedMemory<T> ownedMemory))
             {
-                Memory = MemoryMarshal.AsMemory(readOnlyMemory)
-            };
-            _sequenceStart = new SequencePosition(segment, 0 | MemoryListStartMask);
-            _sequenceEnd = new SequencePosition(segment, readOnlyMemory.Length | MemoryListEndMask);
+                // Its an OwnedMemory so unsafe to unwrap ReadOnlyMemory, so we wrap it instead.
```

Comment:
Why is it unsafe to unwrap?

Reply:
We had this discussion on the original PR but I think it was about unwrapping ReadOnlyMemory to an array? I can't open that PR, it's too big?

Reply:
If can safely unwrap to `OwnedMemory<T>` would the `IMemoryList` interface be needed? Just have a strongly typed version?
```csharp
public class ReadOnlyMemorySegment<T>
{
    public ReadOnlyMemorySegment(ReadOnlyMemory<T> readOnlyMemory)
    {
        Memory = readOnlyMemory;
    }

    public ReadOnlyMemory<T> Memory { get; set; }
    public ReadOnlyMemorySegment<T> Next { get; set; }
    public long RunningIndex { get; set; }
}
```
Then `BufferSegement` could wrap it as a field?

Reply:
Or inherit and sidecar extra data... (but not have the above methods virtual)

Reply:
If we are fine unwrapping to the array I don't think to do so with OwnedMemory is much worse.

Reply:
Then there wouldn't be any need for `IMemoryList` as `ReadOnlyMemorySegment` becomes the dumb chaining class?

Reply:
Then there wouldn't be any need for `IMemoryList` as `ReadOnlyMemorySegment` becomes the POC chaining class?

Reply:
You would still need to know if OwnedMemory is ReadOnlyMemorySegment, so the need for additional branch doesn't go away.

Reply:
Yes but `Memory`, `Next`, and `RunningIndex` all become inlinable simple properties rather than interface dispatch calls

Reply:
We are removing `RunningIndex` and replacing with two new methods that would have to be virtual. I think we considered base class but decided that IMemoryList offers more flexibility. 
/cc @KrzysztofCwalina

Reply:
> You would still need to know if OwnedMemory is ReadOnlyMemorySegment, so the need for additional branch doesn't go away.

It wouldn't be used for a single segment from ROM; as now, as that single segment just becomes OwnedMemory

Reply:
e.g. https://github.com/dotnet/corefx/pull/27455/commits/975575d9b8402ce07e6e16fed594888be7be3bbb

Reply:
Unwrapped

Reply:
Unwrapped and removed `ReadOnlySequenceSegment` so `BufferSegement` is its only real implementation